File: src/org/simbrain/network/neuron_update_rules/IACRule.java
/*
 * Part of Simbrain--a java-based neural network kit
 * Copyright (C) 2005,2007 The Authors.  See http://www.simbrain.net/credits
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */
package org.simbrain.network.neuron_update_rules;

import org.simbrain.network.core.Network.TimeType;
import org.simbrain.network.core.Neuron;
import org.simbrain.network.core.NeuronUpdateRule;
import org.simbrain.network.core.Synapse;
import org.simbrain.network.neuron_update_rules.interfaces.BoundedUpdateRule;
import org.simbrain.network.neuron_update_rules.interfaces.ClippableUpdateRule;
import org.simbrain.network.neuron_update_rules.interfaces.NoisyUpdateRule;
import org.simbrain.util.randomizer.Randomizer;

/**
 * <b>IACNeuron</b> implements an Interactive Activation and Competition neuron.
 */
public class IACRule extends NeuronUpdateRule implements BoundedUpdateRule,
        ClippableUpdateRule, NoisyUpdateRule {

    /** The Default upper bound. */
    private static final double DEFAULT_CEILING = 1.0;

    /** The Default lower bound. */
    private static final double DEFAULT_FLOOR = -.2;

    /** Neuron decay. */
    private double decay = 0.05;

    /** Rest. */
    private double rest = .1;

    /** Noise dialog box. */
    private Randomizer noiseGenerator = new Randomizer();

    /** Add noise to the neuron. */
    private boolean addNoise = false;

    /** Clipping. */
    private boolean clipping = true;

    /** The upper bound of the activity if clipping is used. */
    private double ceiling = DEFAULT_CEILING;

    /** The lower bound of the activity if clipping is used. */
    private double floor = DEFAULT_FLOOR;

    /** Local variables as class variables for a minor performance gain. */
    private double effect, netInput, act;

    /**
     * @return
     * {@inheritDoc}
     */
    public TimeType getTimeType() {
        return TimeType.DISCRETE;
    }

    /**
     * @return
     * {@inheritDoc}
     */
    public IACRule deepCopy() {
        IACRule iac = new IACRule();
        iac.setDecay(getDecay());
        iac.setRest(getRest());
        iac.setClipped(isClipped());
        iac.setUpperBound(getUpperBound());
        iac.setLowerBound(getLowerBound());
        iac.setAddNoise(getAddNoise());
        iac.noiseGenerator = new Randomizer(noiseGenerator);
        return iac;
    }

    @Override
    public void update(Neuron neuron) {

        // Notation and algorithm from McClelland 1981, Proceedings of the third
        // annual cog-sci meeting

        // Sum of the "active excitors" and "active inhibitors"
        netInput = neuron.getInputValue();
        for (Synapse w : neuron.getFanIn()) {
            if (w.getSource().getActivation() > 0) {
                netInput += (w.getStrength() * w.getSource().getActivation());
            }
        }

        // Determine "effect" value.
        effect = 0;
        if (netInput >= 0) {
            effect = (getUpperBound() - neuron.getActivation()) * netInput;
        } else {
            effect = (neuron.getActivation() - getLowerBound()) * netInput;
        }

        // Update activation using Euler integration of main ODE
        act = neuron.getActivation() + neuron.getNetwork().getTimeStep()
                * (effect - decay * (neuron.getActivation() - rest));

        if (addNoise) {
            act += noiseGenerator.getRandom();
        }

        if (clipping) {
            act = clip(act);
        }

        neuron.setBuffer(act);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public double clip(double val) {
        if (val > getUpperBound()) {
            return getUpperBound();
        } else if (val < getLowerBound()) {
            return getLowerBound();
        } else {
            return val;
        }
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void contextualIncrement(Neuron n) {
        double act = n.getActivation();
        if (act >= getUpperBound() && isClipped()) {
            return;
        } else {
            if (isClipped()) {
                act = clip(act + increment);
            } else {
                act = act + increment;
            }
            n.setActivation(act);
            n.getNetwork().fireNeuronChanged(n);
        }
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void contextualDecrement(Neuron n) {
        double act = n.getActivation();
        if (act <= getLowerBound() && isClipped()) {
            return;
        } else {
            if (isClipped()) {
                act = clip(act - increment);
            } else {
                act = act - increment;
            }
            n.setActivation(act);
            n.getNetwork().fireNeuronChanged(n);
        }
    }

    /**
     * @return Returns the decay.
     */
    public double getDecay() {
        return decay;
    }

    /**
     * @param decay The decay to set.
     */
    public void setDecay(final double decay) {
        this.decay = decay;
    }

    /**
     * @return Returns the rest.
     */
    public double getRest() {
        return rest;
    }

    /**
     * @param rest The rest to set.
     */
    public void setRest(final double rest) {
        this.rest = rest;
    }

    /**
     * @return Returns the addNoise.
     */
    public boolean getAddNoise() {
        return addNoise;
    }

    /**
     * @param addNoise The addNoise to set.
     */
    public void setAddNoise(final boolean addNoise) {
        this.addNoise = addNoise;
    }

    /**
     * @return Returns the noiseGenerator.
     */
    public Randomizer getNoiseGenerator() {
        return noiseGenerator;
    }

    /**
     * @param noiseGenerator The noiseGenerator to set.
     */
    public void setNoiseGenerator(final Randomizer noiseGenerator) {
        this.noiseGenerator = noiseGenerator;
    }

    @Override
    public String getDescription() {
        return "IAC";
    }

    @Override
    public double getUpperBound() {
        return ceiling;
    }

    @Override
    public double getLowerBound() {
        return floor;
    }

    @Override
    public void setUpperBound(double ceiling) {
        this.ceiling = ceiling;
    }

    @Override
    public void setLowerBound(double floor) {
        this.floor = floor;
    }

    @Override
    public boolean isClipped() {
        return clipping;
    }

    @Override
    public void setClipped(boolean clipping) {
        this.clipping = clipping;
    }
}

File: src/main/kotlin/org/simbrain/network/updaterules/IACRule.kt
package org.simbrain.network.updaterules

import org.simbrain.network.core.Network
import org.simbrain.network.core.Neuron
import org.simbrain.network.updaterules.interfaces.ClippedUpdateRule
import org.simbrain.network.updaterules.interfaces.NoisyUpdateRule
import org.simbrain.network.util.EmptyMatrixData
import org.simbrain.network.util.EmptyScalarData
import org.simbrain.util.UserParameter
import org.simbrain.util.stats.ProbabilityDistribution
import org.simbrain.util.stats.distributions.UniformRealDistribution

/**
 * **IACNeuron** implements an Interactive Activation and Competition neuron.
 */
class IACRule : NeuronUpdateRule<EmptyScalarData, EmptyMatrixData>(), ClippedUpdateRule, NoisyUpdateRule {
    @UserParameter(
        label = "Decay Rate",
        description = "The rate at which activation decays to its resting value.",
        increment = .1,
        order = 1
    )
    var decay: Double = 0.05

    @UserParameter(
        label = "Rest",
        description = "The resting value which the activation decays to.",
        increment = .1,
        order = 2
    )
    var rest: Double = .1

    /**
     * Noise generator.
     */
    override var noiseGenerator: ProbabilityDistribution = UniformRealDistribution()

    /**
     * Add noise to the neuron.
     */
    override var addNoise: Boolean = false

    /**
     * Clipping.
     */
    override var isClipped: Boolean = true

    /**
     * The upper bound of the activity if clipping is used.
     */
    override var upperBound: Double = DEFAULT_CEILING

    /**
     * The lower bound of the activity if clipping is used.
     */
    override var lowerBound: Double = DEFAULT_FLOOR

    override val timeType: Network.TimeType
        get() = Network.TimeType.DISCRETE

    override fun deepCopy(): IACRule {
        val iac = IACRule()
        iac.decay = decay
        iac.rest = rest
        iac.isClipped = isClipped
        iac.upperBound = upperBound
        iac.lowerBound = lowerBound
        iac.addNoise = addNoise
        iac.noiseGenerator = noiseGenerator.deepCopy()
        return iac
    }

    context(Network)
    override fun apply(neuron: Neuron, data: EmptyScalarData) {
        // Notation and algorithm from McClelland 1981, Proceedings of the third
        // annual cog-sci meeting

        // Sum of the "active excitors" and "active inhibitors"

        var netInput = neuron.input
        for (w in neuron.fanIn) {
            if (w.source.activation > 0) {
                netInput += (w.strength * w.source.activation)
            }
        }

        var effect = 0.0
        effect = if (netInput >= 0) {
            (upperBound - neuron.activation) * netInput
        } else {
            (neuron.activation - lowerBound) * netInput
        }

        // Update activation using Euler integration of main ODE
        var act: Double =
            neuron.activation + timeStep * (effect - decay * (neuron.activation - rest))

        if (addNoise) {
            act += noiseGenerator.sampleDouble()
        }

        neuron.activation = act
        neuron.clip()
    }

    override val name: String
        get() = "IAC"

    companion object {
        /**
         * The Default upper bound.
         */
        private const val DEFAULT_CEILING = 1.0

        /**
         * The Default lower bound.
         */
        private const val DEFAULT_FLOOR = -.2
    }
}